src/emqx_mqueue.erl

```diff
@@ -194,3 +235,46 @@ get_priority_opt(Opts) ->
 %% while the highest 'infinity' is a [{infinity, queue:queue()}]
 get_priority(_Topic, ?NO_PRIORITY_TABLE, _) -> ?LOWEST_PRIORITY;
 get_priority(Topic, PTab, Dp) -> maps:get(Topic, PTab, Dp).
+
+init_counter(?HIGHEST_PRIORITY, Opts) ->
+    Infinity = 1000000,
+    init_counter(Infinity, Opts);
+init_counter(Prio, #shift_opts{multiplier = Mult, base = Base}) ->
+    (Prio + Base) * Mult.
+
+get_shift_opt(Opts) ->
+    Mult = maps:get(shift_multiplier, Opts, 10),
+    Min = case Opts of
+              #{p_table := PTab} ->
+                  case maps:size(PTab) of
+                      0 -> 0;
+                      _ -> lists:min(maps:values(PTab))
+                  end;
+              _ ->
+                  ?LOWEST_PRIORITY
+          end,
+    Base = case Min < 0 of
+               true  -> -Min;
+               false -> 0
```

Comment:
After counter goes down to 0, we call shift function, and then `pqueue:out_p` to get the next element's priority, that is used to initialize the new countdown. So even if we initialize the counter with 0, there is at least one iteration of consuming from this priority. So 0 does not need special treatment.

Reply:
I mean be assertive about: `Min` should never be greater than `0`

Reply:
Sorry, I don't follow. I think it can be greater than 0, if the user configures priorities like this: `#{topic1 => 1, topic2 => 2}`, for example.

Reply:
Ah, I got it mixed up. It’s negative inside pqueue module.

now I understand that this base is to ensure the multiply will not result in a negative value.

Does this mean though, in case of negative priority or zero priority, the lowest priority will always have to shift right after one `out`?